### PR TITLE
Added options to readability-implicit-bool-conversion 

### LINKED ISSUE #36323: Add option to readability-implicit-bool-conversion check to allow implicit conversion *to bool*
|  |  |
| --- | --- |
| Bugzilla Link | [36975](https://llvm.org/bz36975) |
| Version | unspecified |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |

## Extended Description 
From the readability-implicit-bool-conversion check, I'd like to take advantage of these parts:

- boolean expression/literal to integer,
- boolean expression/literal to floating.

But not these:

- integer expression/literal to boolean,
- floating expression/literal to boolean,
- pointer/pointer to member/nullptr/NULL to boolean,

Seems like an `AllowConversionsToBool` option would take care of this. Maybe for completeness/symmetry also a `AllowConversionsFromBool` option. Both defaulting to 0.

### PR DESCRIPTION
**Pull request by 4m4n-x-B4w4ne**
Fixes: #36323 , I added two new options in the clang-tools-extra/clan-tidy/readibility/ImplicitBoolConversionCheck.cpp and header file.
I have also written new test cases to test these new options in test/readibility directory.

### CODE REVIEW

**Comment by EugeneZelenko** [comment]
Please mention options in Release Notes and check documentation.

**Review by HerrCai0907 — COMMENTED** [review]
update release note and doc also

File: clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp:4]: this is old format for options, use new one like in other checks

File: clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp:363]: Wrong, place these things should be actually put into matcher in registerMatchers method, to speed up.

You got 2 calls to addMatcher, one in line 300, other in line 339, put those (with dependences to avoid unused local variables) into if's.

**Review by PiotrZSL — CHANGES_REQUESTED** [review]
- reformat code
- missing release notes entry about new options added
- missing documentation for newly added options
- change should be in registerMatchers method

File: clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp:4]: Okay Thanks I am Working on it.

File: clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp:363]: Okay Thanks I am Working on it.

**Comment by 4m4n-x-B4w4ne** [comment]
> update release note and doc also

I am doing it.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:335]: Please extend existing `readability-implicit-bool-conversion` entry.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:337]: Default belongs to documentation.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:323]: `check by adding the options` should be common for all three options.

**Comment by EugeneZelenko** [comment]
New options are still not described in check documentation.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [reply clang-tools-extra/docs/ReleaseNotes.rst:323]: Sorry, if I was not clear, `check by adding the options` should be added once before list of added options.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst:323]: Thanks I have done it in Release Notes .

Comment (EugeneZelenko) [reply clang-tools-extra/docs/ReleaseNotes.rst:323]: It is not necessary to repeat `check by adding the option` thrice.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Its looking good ?

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:323]: Belongs to previous line.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:324]: Please follow 80-characters limit.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst:323]: Sorry but i havent got this one . Can you please clarify?

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:323]:
```suggestion
   the options:
```

Or `next options:`?

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:324]:
```suggestion
   - `UseUpperCaseLiteralSuffix`: to select the case of the literal suffix in 
```

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:328]:
```suggestion
      implicit conversions to ``bool``.
```

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:330]:
```suggestion
      implicit conversions from ``bool``.
```

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Thanks got it now.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:328]:
```suggestion
      implicit conversions to ``bool``.
```

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp]: Can You please check this? I have done this now as instructed.

File: clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp:291]: Please run Clang-format.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp]: Done Clang-Format .
Thanks.

**Comment by 4m4n-x-B4w4ne** [comment]
> Please mention options in Release Notes and check documentation.

I have documented it can you please review it !

File: clang-tools-extra/docs/clang-tidy/checks/readability/implicit-bool-conversion.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/clang-tidy/checks/readability/implicit-bool-conversion.rst:152]:
```suggestion
   When `true`, the check diagnoses implicit conversions to ``bool``. Default is `true`.
```

As language construct. Please also follow 80 characters limit. Same for other option.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/clang-tidy/checks/readability/implicit-bool-conversion.rst]: Done Thanks.

Can you please also review ImplicitBoolConversionCheck.cpp?

Comment (EugeneZelenko) [reply clang-tools-extra/docs/clang-tidy/checks/readability/implicit-bool-conversion.rst]: Please wait for other reviewers. I deal only with documentation and minor code style issues.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/clang-tidy/checks/readability/implicit-bool-conversion.rst]: Okay Thanks a lot !

File: clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp]: I have done this . Can you please review it?

**Review by PiotrZSL — APPROVED** [review]
Except pointed out findings, looks ok.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp]: Done Thanks.

**Comment by 4m4n-x-B4w4ne** [comment]
Hey @EugeneZelenko  , As everything got reviewed .Is it possible to merge my PR ?

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:324]: Please merge with previous line.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Hey what needs to be done further!

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (5chmidti) [inline clang-tools-extra/docs/ReleaseNotes.rst:324]: Please re-add what was already there to end up with something like: 

```
- Improved :doc:`readability-implicit-bool-conversion
  <clang-tidy/checks/readability/implicit-bool-conversion>` check
  by adding the option `UseUpperCaseLiteralSuffix` to select the
  case of the literal suffix in fixes and fixing false positive for implicit
  conversion of comparison result in C23, and by adding the
  options `CheckConversionsToBool` and
  `CheckConversionsFromBool` to configure which conversions 
  involving ``bool`` are allowed.
```

File: clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp:298]: The PR doesn't build because you have a typo: `CheckConversionToBool` -> `CheckConversionsToBool`, the same applies to `CheckConversionFromBool` below.

File: clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp
Comment (5chmidti) [inline clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp:1]: You have added `-check-suffix` options to each of your `RUN` lines, but none of your `CHECK-MESSAGES` are actually using it. E.g.,: 
```C++
// CHECK-MESSAGES-TO-BOOL-FALSE
```

Although, instead of using the suffixes to say `FALSE`, you could swap them around to say that the option is enabled, so `-check-suffix:TO-BOOL`. Then, the version with both enabled becomes `-check-suffix:TO-BOOL,FROM-BOOL`.

File: clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/clang-tidy/readability/ImplicitBoolConversionCheck.cpp]: Done Thanks. Sorry for my carelessness.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

File: clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp
Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp:1]: Thanks ,I am working on it and also on clean PR build.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp:1]: I have wrote a basic-level test similar to other checks and also have done a clean PR build.
Can you please review it?

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/test/clang-tidy/checkers/readability/implicit-bool-conversion-check.cpp:1]: I have written check file for all permutations of configuration of the options and also for the more comprehensive test cases as instructed with clean PR build.
Done Thanks.
@5chmidti  Can you please review it?

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:325]:
```suggestion
  C23, and by adding the option `CheckConversionsToBool` or 
```

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Hey, What needs to be done further? Can you please help me out?

**Review by HerrCai0907 — APPROVED** [review]
LGTM. If you do not have write access, please ping me. we can help you to do merge.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (EugeneZelenko) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Please fix Clang-format complains.

Comment (4m4n-x-B4w4ne) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done Thanks.

**Comment by 4m4n-x-B4w4ne** [comment]
@HerrCai0907 , Thanks I dont have any write permissions. Can you please help me to merge?

**Comment by chapuni** [comment]
Excuse me, I've reverted this.

**Comment by 4m4n-x-B4w4ne** [comment]
> Excuse me, I've reverted this.

Can you help on what thing , It went wrong and like ,What shall I need to do to make it done?

**Comment by kaddkaka** [comment]
I'm interested in this configuration. 

This was reverted after tests broke. What's the status on this now?

@chapuni do you have some inputs on how to do test verification before trying to merge this again?

**Comment by chapuni** [comment]
@kaddkaka IIRC the issue was "char is unsigned". It could be reproducible with adding default target as failing targets above.

**Pull request merged**
